source/common/ssl/context_impl.cc

```diff
@@ -469,12 +480,79 @@ ServerContextImpl::ServerContextImpl(ContextManagerImpl& parent, Stats::Scope& s
     RELEASE_ASSERT(rc == 1);
   }
 
+  // Hash configured SNIs for this context, so that sessions cannot be resumed across different
+  // filter chains, even when using the same server certificate.
+  for (const auto& name : server_names_) {
+    rc = EVP_DigestUpdate(&md, name.data(), name.size());
+    RELEASE_ASSERT(rc == 1);
+  }
+
   rc = EVP_DigestFinal(&md, session_context_buf, &session_context_len);
   RELEASE_ASSERT(rc == 1);
   rc = SSL_CTX_set_session_id_context(ctx_.get(), session_context_buf, session_context_len);
   RELEASE_ASSERT(rc == 1);
 }
 
+ssl_select_cert_result_t
+ServerContextImpl::processClientHello(const SSL_CLIENT_HELLO* client_hello) {
+  std::string server_name = EMPTY_STRING;
+  const uint8_t* data;
+  size_t len;
+
+  if (SSL_early_callback_ctx_extension_get(client_hello, TLSEXT_TYPE_server_name, &data, &len)) {
+    // Based on BoringSSL's ext_sni_parse_clienthello().
```

Comment:
Any reason we can't use the boring ssl function here? Is it worth putting in a TODO here to replace with a boringssl helper? cc @davidben

Reply:
Few reasons:
1. `ext_sni_parse_clienthello()` isn't exported and early callback (which we need to be able to set different protocol versions across different filter chains) is called so early that SNI isn't parsed yet and it's not available via `SSL_set_tlsext_host_name()`.
2. I'm going to need a few more changes for ClientHello sniffing (for #1843), so I'm going to try to upstream everything that's needed once we have a complete picture.
3. We're using `chromium-stable` branch of BoringSSL, so anything we add to BoringSSL wouldn't be available to use for the next month or two anyway.

Reply:
OK SGTM. Maybe just add a TODO to circle back and upstream.

Reply:
Done.